mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp

```diff
@@ -0,0 +1,220 @@
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/FunctionSupport.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/Utils.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "normalize-memrefs"
+
+using namespace mlir;
+
+namespace {
+
+/// All interprocedural memrefs with non-trivial layout maps are converted to
+/// ones with trivial identity layout ones.
+
+// Input :-
+// #pad = affine_map<(i) -> (i floordiv 4, i mod 4)>
+// func @matmul(%A: memref<16xf64, #pad>, %B: index, %C: memref<16xf64>) ->
+// (memref<16xf64, #pad>) {
+//   affine.for %arg3 = 0 to 16 {
+//         %a = affine.load %A[%arg3] : memref<16xf64, #pad>
+//         %p = mulf %a, %a : f64
+//         affine.store %p, %A[%arg3] : memref<16xf64, #pad>
+//   }
+//   %c = alloc() : memref<16xf64, #pad>
+//   %d = affine.load %c[0] : memref<16xf64, #pad>
+//   return %A: memref<16xf64, #pad>
+// }
+
+// Output :-
+// module {
+//   func @matmul(%arg0: memref<4x4xf64>, %arg1: index, %arg2: memref<16xf64>)
+//   -> memref<4x4xf64> {
+//     affine.for %arg3 = 0 to 16 {
+//       %2 = affine.load %arg0[%arg3 floordiv 4, %arg3 mod 4] : memref<4x4xf64>
+//       %3 = mulf %2, %2 : f64
+//       affine.store %3, %arg0[%arg3 floordiv 4, %arg3 mod 4] : memref<4x4xf64>
+//     }
+//     %0 = alloc() : memref<16xf64, #map0>
+//     %1 = affine.load %0[0] : memref<16xf64, #map0>
+//     return %arg0 : memref<4x4xf64>
+//   }
+// }
+
+struct NormalizeMemRefs : public NormalizeMemRefsBase<NormalizeMemRefs> {
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+std::unique_ptr<OperationPass<ModuleOp>> mlir::createNormalizeMemRefsPass() {
+  return std::make_unique<NormalizeMemRefs>();
+}
+
+void NormalizeMemRefs::runOnOperation() {
+  auto module = getOperation();
+
+  module.walk([&](FuncOp funcOp) {
+    OpBuilder b(funcOp);
+
+    FunctionType ft = funcOp.getType();
+    SmallVector<Type, 8> inputs;  // Holds the arguments of function.
+    SmallVector<Type, 4> results; // Holds the results of function.
+
+    // Populating results with function's initial result (type) as
+    // this will be used to modify and set function's signature later.
+    for (auto retIndex : llvm::seq<unsigned>(0, funcOp.getNumResults())) {
+      results.push_back(ft.getResult(retIndex));
+    }
+
+    // Walk over each argument of a function to perform memref normalization (if
+    // any).
+    for (auto argIndex : llvm::seq<unsigned>(0, funcOp.getNumArguments())) {
+
+      // Check whether argument is of MemRef type.
+      // TODO(avarmapml): What about unranked memrefs?
+      if (funcOp.getArgument(argIndex).getType().getKind() ==
+          StandardTypes::MemRef) {
```

Comment:
Use early exit to avoid extra indentation for all the following code and thus making it simpler / easier to read, i.e., 
```
if (!funcOp.getArgument(argIndex).getType().isa<MemRefType>())
  continue;
// The rest of the code here will have one indentation less.
```